ibc-testkit/src/testapp/ibc/core/core_ctx.rs

```diff
@@ -298,46 +268,449 @@ impl ValidationContext for MockContext {
+        sequences: impl ExactSizeIterator<Item = Sequence>,
+    ) -> Result<Vec<Sequence>, ContextError> {
+        // QUESTION. Currently only works for unordered channels; ordered channels
+        // don't use receipts. However, ibc-go does it this way. Investigate if
+        // this query only ever makes sense on unordered channels.
+
+        Ok(sequences
+            .into_iter()
+            .map(|seq| ReceiptPath::new(&channel_end_path.0, &channel_end_path.1, seq))
+            .filter(|receipt_path| {
+                self.packet_receipt_store
+                    .get(StoreHeight::Pending, receipt_path)
+                    .is_none()
+            })
+            .map(|receipts_path| receipts_path.sequence)
+            .collect())
+    }
+
+    /// UnreceivedAcks returns all the unreceived IBC acknowledgements associated with a channel and sequences.
+    /// Returns all the unreceived acks if sequences is empty.
+    fn unreceived_acks(
+        &self,
+        channel_end_path: &ChannelEndPath,
+        sequences: impl ExactSizeIterator<Item = Sequence>,
+    ) -> Result<Vec<Sequence>, ContextError> {
+        let collected_paths: Vec<_> = if sequences.len() == 0 {
+            // if sequences is empty, return all the acks
+            let commitment_path_prefix = format!(
+                "commitments/ports/{}/channels/{}/sequences",
+                channel_end_path.0, channel_end_path.1
+            )
+            .try_into()
+            .map_err(|_| PacketError::Other {
+                description: "Invalid commitment path".into(),
+            })?;
+
+            self.packet_commitment_store
+                .get_keys(&commitment_path_prefix)
+                .into_iter()
+                .filter_map(|path| {
+                    if let Ok(Path::Commitment(commitment_path)) = path.try_into() {
+                        Some(commitment_path)
+                    } else {
+                        None
+                    }
+                })
+                .collect()
+        } else {
+            sequences
+                .into_iter()
+                .map(|seq| CommitmentPath::new(&channel_end_path.0, &channel_end_path.1, seq))
+                .collect()
+        };
+
+        Ok(collected_paths
+            .into_iter()
+            .filter(|commitment_path: &CommitmentPath| -> bool {
+                self.packet_commitment_store
+                    .get(StoreHeight::Pending, commitment_path)
+                    .is_some()
+            })
+            .map(|commitment_path| commitment_path.sequence)
+            .collect())
+    }
+}
+
+impl<S> ExecutionContext for MockIbcStore<S>
+where
+    S: ProvableStore + Debug,
+{
     type E = Self;
 
     fn get_client_execution_context(&mut self) -> &mut Self::E {
         self
     }
 
+    /// Called upon client creation.
+    /// Increases the counter which keeps track of how many clients have been created.
+    /// Should never fail.
```

Comment:
It seems a bit strange that the docstring here specifically points out that this method should never fail. Is there a particular reason for that? Nothing in the function's body or its signature would indicate that it should never fail. If that's an important point, then it should perhaps be baked more into the function's implementation, for example by not having it return a `Result` or something like that.

Reply:
It is not written by me. But I agree this `Should never fail` is not very helpful. I think, this just means - in a correct blockchain - this call never fails. But in that case, there are other methods here that deserves the same comment. I will remove it.

Reply:
d8fe3d4b